File: Python/algorithmic_heights/21_sq.py
# Square in a Graph
import sys

from collections import defaultdict


class Graph:
    def __init__(self, nr_nodes, edge_list):
        self._nr_nodes = nr_nodes
        self._adjacency_list = self._build_adjacency_list(edge_list)

    @staticmethod
    def _build_adjacency_list(edge_list):
        adjacency_list = defaultdict(set)
        for a, b in edge_list:
            adjacency_list[a].add(b)
            adjacency_list[b].add(a)
        return dict(adjacency_list)

    @property
    def nr_nodes(self):
        return self._nr_nodes

    def neighbours(self, node):
        return self._adjacency_list.get(node, set())


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        a, b = convert_to_intlist(next(reader))
        edges.append((a, b))
    return edges


def read_graphs(reader, n):
    graphs = []
    for _ in range(n):
        _ = next(reader)
        nr_nodes, nr_edges = convert_to_intlist(next(reader))
        edge_list = read_edges(reader, nr_edges)
        graph = Graph(nr_nodes, edge_list)
        graphs.append(graph)
    return graphs


def has_cycle_of_length_four(graph):
    for node1 in range(1, graph.nr_nodes+1):
        if (neighbours1 := graph.neighbours(node1)):
            for node2 in range(node1+1, graph.nr_nodes+1):
                neighbours2 = graph.neighbours(node2)
                if len(neighbours1 & neighbours2) >= 2:
                    return True
    return False


def have_graphs_squares(graphs):
    return [has_cycle_of_length_four(graph) for graph in graphs]


def main():
    reader = sys.stdin
    nr_examples = int(next(reader))
    graphs = read_graphs(reader, nr_examples)
    result = have_graphs_squares(graphs)
    print(' '.join(map(lambda x: '1' if x else '-1', result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/21_sq.hs
import Control.Monad (replicateM)
import Data.IntMap (IntMap)
import Data.IntSet (IntSet)
import qualified Data.IntMap as M
import qualified Data.IntSet as S

data Edge = Edge Int Int
data Graph = Graph { nrNodes :: Int, adjacencyList :: IntMap IntSet }

convertToIntList :: String -> [Int]
convertToIntList = map read . words


readInputLine :: String -> (Int, Int)
readInputLine line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


createGraph :: Int -> [Edge] -> Graph
createGraph nrNodes edges =
    let addEdge adjacencyList (Edge a b) =
            let update Nothing = Just (S.singleton b)
                update (Just ns) = Just (S.insert b ns)
            in M.alter update a adjacencyList
        addEdgeBothWays adjacencyList edge @ (Edge a b) =
            addEdge (addEdge adjacencyList edge) (Edge b a)
        adjacencyList = foldl addEdgeBothWays M.empty edges
    in Graph { nrNodes = nrNodes, adjacencyList = adjacencyList }


readGraphs :: Int -> IO [Graph]
readGraphs n =
    let readEdges nrEdges = do
            edges <- replicateM nrEdges getLine
            return $ map (\line -> let (a, b) = readInputLine line in Edge a b) edges
        readGraph = do
            _ <- getLine
            (nrNodes, nrEdges) <- readInputLine <$> getLine
            edges <- readEdges nrEdges
            return $ createGraph nrNodes edges
    in mapM (const readGraph) [1..n]


getNeighbors :: Graph -> Int -> IntSet
getNeighbors (Graph _ adjacencyList) node = M.findWithDefault S.empty node adjacencyList


hasCycleOfLengthFour :: Graph -> Bool
hasCycleOfLengthFour graph @ (Graph nrNodes _) =
    let predicate node1 =
            let ns1 = getNeighbors graph node1
                twoCommonNeighbors = any(\node2 ->
                    let ns2 = getNeighbors graph node2
                    in S.size (S.intersection ns2 ns1) >= 2) [(node1+1)..nrNodes]
            in not (S.null ns1) && twoCommonNeighbors
    in any predicate [1..nrNodes]


main :: IO ()
main = do
    n <- readLn
    graphs <- readGraphs n
    let results = map hasCycleOfLengthFour graphs
    putStrLn $ unwords $ map (\res -> if res then "1" else "-1") results
